tests/virt_autotest/sev_snp_guest_verification.pm

```diff
@@ -0,0 +1,1270 @@
+
+Without the MSR module, some SEV-SNP diagnostics and attestation functions may fail,
+but the basic SEV-SNP functionality should still work (see bsc#1237858).
+
+=cut
+
+sub setup_msr_module {
+    my $self = shift;
+
+    # Reference bug for MSR module issues with SEV-SNP
+    record_soft_failure("bsc#1237858 - Failed to load msr module may cause SNP attestation to fail");
+
+    # Try to load the MSR module using script_run for better error handling
+    record_info('MSR Module', "Checking and loading MSR module for SEV-SNP verification");
+
+    # First check if module is already loaded
+    if (script_run("lsmod | grep -q '^msr'") == 0) {
+        record_info('MSR Module', "MSR module is already loaded in kernel");
+        return $self;
+    }
+
+    # Module not loaded, try to load it
+    if (script_run("modprobe msr") != 0) {
+        record_info('MSR Module Load Failed', "Failed to load MSR module, some SEV-SNP diagnostics and attestation may fail", result => 'softfail');
+    } else {
+        record_info('MSR Module Loaded', "Successfully loaded MSR module for SEV-SNP verification");
+    }
+
+    return $self;
+}
+
+=head2 configure_sev_snp_kernel_parameters
+
+  configure_sev_snp_kernel_parameters($self)
+
+Configure the necessary kernel parameters for SEV-SNP support using bootloader_setup library.
+This function checks for the required parameters, adds any missing ones,
+updates GRUB configuration, and verifies the changes after reboot.
+
+Required kernel parameters for SEV-SNP:
+- amd_iommu=on: Enables AMD IOMMU support (required for memory encryption)
+- iommu=nopt: Enables IOMMU no pass-through mode
+- mem_encrypt=on: Enables memory encryption features
+- kvm_amd.sev=1: Enables AMD SEV (Secure Encrypted Virtualization) base functionality
+- kvm-amd.sev_snp=1: Enables AMD SEV-SNP (Secure Nested Paging) functionality
+
+=cut
+
+sub configure_sev_snp_kernel_parameters {
+    my $self = shift;
+
+    # Required kernel parameters for SEV-SNP
+    my %required_params = (
+        'amd_iommu=on' => 'AMD IOMMU support',
+        'iommu=nopt' => 'IOMMU no pass-through mode',
+        'mem_encrypt=on' => 'Memory encryption',
+        'kvm_amd.sev=1' => 'SEV base functionality',
+        'kvm-amd.sev_snp=1' => 'SEV-SNP functionality'
+    );
+
+    # Check current kernel parameters
+    my $cmdline = script_output("cat /proc/cmdline");
+    record_info('Kernel cmdline', $cmdline);
+
+    # Find missing parameters
+    my @missing;
+    foreach my $param (keys %required_params) {
+        push @missing, $param unless $cmdline =~ /\b$param\b/;
+    }
+
+    # If all parameters are present, return success
+    if (!@missing) {
+        record_info('SEV-SNP boot parameters', 'All required parameters present');
+        return 1;
+    }
+
+    # Log missing parameters
+    my $missing_list = join(", ", @missing);
+    record_info('Missing SEV-SNP parameters', "Missing parameters: $missing_list", result => 'softfail');
```

Comment:
Why would it be softfail? From code, it seems there is no earlier host kernel parameter preparation for snp, and by default it should not be directly ready for snp either.

Reply:
I removed result also.